File: app/src/main/java/ru/yanus171/android/autobrightness/NodeList.kt
package ru.yanus171.android.autobrightness

import android.preference.PreferenceManager
import android.text.TextUtils
import androidx.core.content.edit

private const val NODE_LIST_PREF = "node_list5"
private const val NODE_SEP = "|"
private const val NODE_FIELD_SEP = ";"
const val NODE_COUNT = 20
const val MAX_BRIGHTNESS = 255
const val MIN_BRIGHTNESS = 0
const val SENSOR_STEP_COEFF = 1.5
class NodeList(private val mMaxSensorValue: Int) {
    class Node {
        constructor(sensorValue: Int, brightness: Int) {
            mBrightness = brightness
            mSensorValue = sensorValue
        }

        constructor(s: String) {
            val list = s.split(NODE_FIELD_SEP)
            mBrightness = list[0].toInt()
            mSensorValue = list[1].toInt()
        }
        fun getString(): String {
            return "s=$mSensorValue, b=$mBrightness"
        }

        fun save(): String {
            return "$mBrightness$NODE_FIELD_SEP$mSensorValue"
        }

        internal var mBrightness: Int = 0
        var mSensorValue: Int = 0


    }

    private var mList: MutableList<Node> = mutableListOf()

    init {
        val settings = PreferenceManager.getDefaultSharedPreferences(MainApplication.context)
        var s = settings.getString(NODE_LIST_PREF, null)
        if (s == null)
            createDefaultList()
        else
            for (item in s.split(NODE_SEP))
                mList.add(Node(item))
    }

    private fun createDefaultList() {
        val bStep = MAX_BRIGHTNESS / NODE_COUNT
        var b = 0
        var s = 10
        mList.clear()
        while (s < mMaxSensorValue) {
            s = (s * 1.5).toInt()
            b += bStep
            mList += Node(s, b)
        }
    }

    fun getBrightness(sensorValue: Int): Int {
        val node = getNodeFor( sensorValue )
        if ( node != null )
            return node.mBrightness
        return MAX_BRIGHTNESS
    }
    private fun getNodeFor(sensorValue: Int): Node? {
        for (node in mList)
            if ( sensorValue < node.mSensorValue )
                return node
        return null
    }


    fun addNodesTo(maxSensorValue: Int ) {
        var sensorValue = mList.last().mSensorValue
        val brightness = mList.last().mBrightness
        while ( sensorValue < maxSensorValue ) {
            sensorValue = (sensorValue * SENSOR_STEP_COEFF).toInt()
            mList += Node( sensorValue, brightness )
        }
    }

    fun set(sensorValue: Int, newBrightness: Int) {
        if ( sensorValue > mList.last().mSensorValue )
            addNodesTo( sensorValue )
        for (it in mList)
            if (sensorValue < it.mSensorValue) {
                it.mBrightness = newBrightness
                makePrevNonesNotBrighterThen( it )
                makeNextNonesNotDarkerThen( it )
                break
            }
        save()
    }

    private fun makePrevNonesNotBrighterThen( node: Node ) {
        for (it in mList.reversed())
            if ( it.mSensorValue < node.mSensorValue && it.mBrightness > node.mBrightness )
                it.mBrightness = node.mBrightness
    }
    private fun makeNextNonesNotDarkerThen( node: Node ) {
        for (it in mList)
            if ( it.mSensorValue > node.mSensorValue && it.mBrightness < node.mBrightness )
                it.mBrightness = node.mBrightness
    }
    fun getString( sensorValue: Int ): String {
        val currentNode = getNodeFor( sensorValue )
        var result = mutableListOf<String>()
        for (node in mList) {
            val currentText = if ( currentNode != null && currentNode.mSensorValue == node.mSensorValue ) "* " else ""
            result += (currentText + node.getString())
        }
        return TextUtils.join("\n", result)
    }

    fun save() {
        var result = mutableListOf<String>()
        for (node in mList)
            result += node.save()
        val settings = PreferenceManager.getDefaultSharedPreferences(MainApplication.context)
        settings.edit() { putString(NODE_LIST_PREF, TextUtils.join(NODE_SEP, result)) }
    }
}

File: lib/models/node_list.dart
import 'package:shared_preferences/shared_preferences.dart';

class Node {
  int brightness;
  int sensorValue;

  Node(this.sensorValue, this.brightness);

  Node.fromString(String s)
      : brightness = int.parse(s.split(';')[0]),
        sensorValue = int.parse(s.split(';')[1]);

  String save() => '$brightness;$sensorValue';
  String getString() => 's=$sensorValue, b=$brightness';
}

class NodeList {
  static const String nodeListPref = 'node_list5';
  static const int maxBrightness = 255;
  static const int minBrightness = 0;
  static const int nodeCount = 20;

  List<Node> _list = [];

  // Legacy constructor keeps background async loading to avoid breaking callers,
  // but prefer using `NodeList.load()` for explicit, awaited initialization.
  NodeList() {
    _loadFromPreferences();
  }

  NodeList._internal();

  static Future<NodeList> load() async {
    final nl = NodeList._internal();
    await nl._loadFromPreferences();
    return nl;
  }

  int getBrightness(int sensorValue) {
    final node = _getNodeFor(sensorValue);
    return node?.brightness ?? maxBrightness;
  }

  Node? _getNodeFor(int sensorValue) {
    for (final node in _list) {
      if (sensorValue < node.sensorValue) {
        return node;
      }
    }
    return null;
  }

  void set(int sensorValue, int newBrightness) {
    for (final node in _list) {
      if (sensorValue < node.sensorValue) {
        node.brightness = newBrightness;
        _enforceConstraints(node);
        break;
      }
    }
    _save();
  }

  void _enforceConstraints(Node node) {
    // makePrevNonesNotBrighterThen
    for (final it in _list.reversed) {
      if (it.sensorValue < node.sensorValue && it.brightness > node.brightness) {
        it.brightness = node.brightness;
      }
    }
    // makeNextNonesNotDarkerThen
    for (final it in _list) {
      if (it.sensorValue > node.sensorValue && it.brightness < node.brightness) {
        it.brightness = node.brightness;
      }
    }
  }

  String getString(int sensorValue) {
    final currentNode = _getNodeFor(sensorValue);
    return _list.map((node) {
      final current = (currentNode?.sensorValue == node.sensorValue) ? '* ' : '';
      return current + node.getString();
    }).join('\n');
  }

  Future<void> _loadFromPreferences() async {
    final prefs = await SharedPreferences.getInstance();
    final nodeListString = prefs.getString(nodeListPref);

    if (nodeListString == null) {
      _createDefaultList();
    } else {
      _list = nodeListString.split('|').map((item) => Node.fromString(item)).toList();
    }
  }

  void _createDefaultList() {
    _list.clear();
    final bStep = maxBrightness ~/ nodeCount;
    int b = 0;
    int s = 10;

    while (s < 1000) { // Assuming max sensor value
      s = (s * 1.5).toInt();
      b += bStep;
      _list.add(Node(s, b));
    }
  }

  Future<void> _save() async {
    final prefs = await SharedPreferences.getInstance();
    final nodeListString = _list.map((node) => node.save()).join('|');
    await prefs.setString(nodeListPref, nodeListString);
  }
}
